colorizer_data/utils.py

```diff
@@ -302,3 +304,158 @@ def copy_remote_or_local_file(src_path: str, dst_path: str) -> None:
         if not os.path.exists(src_path):
             raise FileNotFoundError(f"Backdrop image '{src_path}' does not exist.")
         shutil.copyfile(src_path, dst_path)
+
+
+def convert_string_array_to_categorical_feature(
+    data: np.ndarray, info: FeatureInfo
+) -> Tuple[np.ndarray, FeatureInfo]:
+    """
+    Parses a string feature into the expected categorical feature format. Returns an updated copy of
+    `info` where `info.categories` is a list of unique strings in order of initial appearance, and a new data
+    integer array which indexes into `info.categories`.
+
+    Example:
+    ```
+    info = FeatureInfo(type=FeatureType.CATEGORICAL, categories=[])
+    data = np.array(["A", "B", "C", "A", "D"], dtype=str)
+
+    new_data, new_info = convert_string_array_to_categorical_feature(data, info)
+    print(new_data)  # [0, 1, 2, 0, 3]
+    print(new_info.categories)  # ["A", "B", "C", "D"]
+    ```
+    """
+    new_info = info.clone()
+    categories, indexed_data = np.unique(data.astype(str), return_inverse=True)
+    new_info.categories = categories.tolist()
+    new_info.type = FeatureType.CATEGORICAL
+    return (indexed_data, new_info)
+
+
+def infer_feature_type(data: np.ndarray, info: FeatureInfo) -> FeatureType:
+    """
+    Infer a concrete feature type from possibly unknown feature data and info types.
+
+    Args:
+        data (np.ndarray): The feature's data array.
+        info (FeatureInfo): The feature's metadata.
+
+    Returns:
+        - If `info.type` is concrete (not `FeatureType.INDETERMINATE`), returns the type.
+        - Otherwise, returns either `CATEGORICAL`, `DISCRETE`, or `CONTINUOUS` based on the type of the data array
+        and other `info` metadata.
+    """
+    if info.type != FeatureType.INDETERMINATE:
+        return info.type
+
+    # See https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html
+    kind = data.dtype.kind
+    if info.categories is not None and len(info.categories) > 0:
```

Comment:
If categories is empty, it'll override it and try and automatically infer the categories from the data.

Reply:
Mm, it depends on the dtype of the `data` array. If `data` is a numeric array, it'll return either the `DISCRETE` or `CONTINUOUS` type.

If `data` is a string or other non-numeric data type, it will return `CATEGORICAL`, since it can infer the categories array.

Reply:
(By default, `info.categories` is `None`, so hopefully no one will assign an empty array to it.)